servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpMessageDiscardWatchdogServiceFilter.java

```diff
@@ -0,0 +1,127 @@
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * Filter which tracks HTTP messages sent by the service, so it can be freed if discarded in the pipeline.
+ *
+ * @see HttpMessageDiscardCleanerServiceLifecycleObserver
+ */
+final class HttpMessageDiscardWatchdogServiceFilter implements StreamingHttpServiceFilterFactory {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpMessageDiscardWatchdogServiceFilter.class);
+
+    /**
+     * Instance of {@link HttpMessageDiscardWatchdogServiceFilter}.
+     */
+    static final StreamingHttpServiceFilterFactory INSTANCE = new HttpMessageDiscardWatchdogServiceFilter();
+
+    @SuppressWarnings("rawtypes")
+    static final ContextMap.Key<AtomicReference> MESSAGE_PUBLISHER_KEY = ContextMap.Key
+            .newKey("io.servicetalk.http.netty.messagePublisher", AtomicReference.class);
+
+    private HttpMessageDiscardWatchdogServiceFilter() {
+        // Singleton
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(final StreamingHttpService service) {
+
+        return new StreamingHttpServiceFilter(service) {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                return delegate()
+                        .handle(ctx, request, responseFactory)
+                        .map(response -> {
+                            // always write the buffer publisher into the request context. When a downstream subscriber
+                            // arrives, mark the message as subscribed explicitly (having a message present and no
+                            // subscription is an indicator that it must be freed later on).
+                            final AtomicReference<?> previous = request.context()
+                                    .put(MESSAGE_PUBLISHER_KEY, new AtomicReference<>(response.messageBody()));
+                            if (previous != null) {
+                                // If a previous message exists, the response publisher got resubscribed to (i.e.
+                                // during a retry) and so also needs to be cleaned up.
+                                Publisher<?> message = (Publisher<?>) previous.get();
+                                if (message != null) {
+                                    LOGGER.debug("Cleaning up HTTP response message which has been resubscribed to - " +
```

Comment:
Recommended rewording above and promoted to `warn` level.

In this case, we probably don't need to worry about "logging once" because (1) it's not expected that users will ever have retries configured and (2) retries are not expected to happen all the time to flood logs. wdyt?

Reply:
Agreed - if we see it spamming we can always bring in the check here too.